src/vumi2/connectors.py

```diff
@@ -73,28 +77,49 @@ async def start(self) -> None:
         await channel.queue_declare(self.queue_name, durable=self.durable)
         await channel.queue_bind(self.queue_name, self.exchange_name, self.queue_name)
         await channel.basic_consume(self.queue_message, queue_name=self.queue_name)
-        for _ in range(self.concurrency):
-            self.nursery.start_soon(self.consume_message)
+        async with self.receive_channel:
+            for _ in range(self.concurrency):
+                self._active_consumers += 1
+                receive_channel = self.receive_channel.clone()
+                self.nursery.start_soon(self.consume_messages, receive_channel)
+
+    async def aclose(self):
+        if not self._closing:
+            self._closing = True
+            self.send_channel.close()
+        await self._closed.wait()
+        if self.channel.is_open:
+            await self.channel.close()
 
     async def queue_message(
         self, channel: Channel, body: bytes, envelope: Envelope, properties: Properties
     ) -> None:
-        await self.send_channel.send((channel, body, envelope, properties))
-
-    async def consume_message(self) -> None:
-        async for channel, body, envelope, _ in self.receive_channel:
-            try:
-                msg = self.message_class.deserialise(json.loads(body))
-                await maybe_awaitable(self.callback(msg))
-                await channel.basic_client_ack(envelope.delivery_tag)
-            except Exception as e:
-                logger.exception(e)
-                await channel.basic_client_nack(
-                    delivery_tag=envelope.delivery_tag, requeue=False
-                )
-
-
-class Publisher:
+        if not self._closing:
+            await self.send_channel.send((channel, body, envelope, properties))
```

Comment:
Should we raise something if we are closing? I'm worried that something gets marked as completed successfully here, but nothing was actually queued.

Reply:
If we never send the message to a handler, it never gets acked/nacked and should thus be requeued by the broker when we close the AMQP channel. (Unless we've told the broker to treat the message as delivered as soon as it reaches a consumer, in which case it's already too late to do anything here anyway.)

Reply:
We really should have tests for this sort of thing, though. I'll see about writing some.